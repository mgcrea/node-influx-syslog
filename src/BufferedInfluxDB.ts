import {InfluxDB, IPoint, IWriteOptions} from 'influx';
import debounce from 'lodash.debounce';

import withUniqueTimestamps from 'src/utils/withUniqueTimestamps';

type BufferedInfluxDBOptions = {wait?: number; maxWait?: number};

class BufferedInfluxDB extends InfluxDB {
  private __stackedPoints: IPoint[];
  constructor(options: any, {wait = 1000}: BufferedInfluxDBOptions = {}) {
    super(options);
    this.__stackedPoints = [];
    this.__debouncedFlushPoints = debounce(this.__flushPoints, wait, {maxWait: wait});
  }
  async writePoints(points: IPoint[], options?: IWriteOptions): Promise<void> {
    this.__stackedPoints.push(...points);
    this.__debouncedFlushPoints(options);
  }
  async __flushPoints(options?: IWriteOptions): Promise<void> {
    const stackedPointsSoFar = this.__stackedPoints.splice(0, this.__stackedPoints.length);
    const uniqueStackedPoints = withUniqueTimestamps(stackedPointsSoFar);
    try {
      return await super.writePoints(uniqueStackedPoints, options);
    } catch (err) {
      // Restore stack on error
      this.__stackedPoints.unshift(...stackedPointsSoFar);
      throw err;
    }
  }
  // typings
  async __writePoints(_points: IPoint[], _options?: IWriteOptions): Promise<void> {}
  async __debouncedFlushPoints(_options?: IWriteOptions): Promise<void> {}
}

export default BufferedInfluxDB;
