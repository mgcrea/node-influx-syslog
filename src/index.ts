import {createLogger, createSyslogStream} from './adapters/bunyan';
import {SYSLOG_SCHEMA} from './adapters/syslog';
import BufferedInfluxDB from './utils/BufferedInfluxDB';

export {SYSLOG_SCHEMA, BufferedInfluxDB, createSyslogStream};

export default createLogger;
