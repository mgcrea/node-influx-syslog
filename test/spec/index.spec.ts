import * as influxSyslog from '../../src';

describe('influxSyslog', () => {
  it('should expose a stable API', () => {
    expect(influxSyslog).toBeDefined();
    expect(influxSyslog).toMatchSnapshot();
  });
});
