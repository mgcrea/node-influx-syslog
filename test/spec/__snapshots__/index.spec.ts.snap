// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`influxSyslog should expose a stable API 1`] = `
Object {
  "BufferedInfluxDB": [Function],
  "SYSLOG_SCHEMA": Object {
    "fields": Object {
      "facility_code": 1,
      "message": 2,
      "procid": 2,
      "severity_code": 1,
      "timestamp": 1,
      "version": 1,
    },
    "measurement": "syslog",
    "tags": Array [
      "severity",
      "facility",
      "host",
      "hostname",
      "appname",
    ],
  },
  "__esModule": true,
  "createSyslogStream": [Function],
  "default": [Function],
}
`;
